### PR TITLE
[Refactor] Deprecating MetaTensor

### PR DESCRIPTION
**Pull request by vmoens**
## Description

Less code is always better code.

Not marking this as bc breaking as meta-tensors was a private feature.

#### This branch

All benchmarks are faster:

```
TensorDict({}, [3, 4]) 0.05991072800000019
TensorDict({'a': a, 'b': b}, [3, 4]) 0.11035974999999976
TensorDict({'a': a, ('b', 'b1'): b}, [3, 4]) 0.2848203609999995
TensorDict({'a': a, 'b': {'b1': b}}, [3, 4]) 0.2066304670000001
tdc = td.clone() 0.2708355080000002
tdc = td.clone();tdc['c'] = c 0.380355711
tdc = td.clone();tdc.set('c', c) 0.3110597049999999
tdc = td.clone();tdc.share_memory_() 2.1743871629999987
tdc = td.clone();tdc.update(td2) 0.3615297230000003
tdc = td.clone();tdc.update(td2) 0.560022472
tdc = td.clone();tdc['b', 'b1'] = b 0.44018999600000086
tdc = td.clone();tdc['c', 'c', 'c'] = c 0.7462314710000015
```

#### main

```
TensorDict({}, [3, 4]) 0.08259503499999976
TensorDict({'a': a, 'b': b}, [3, 4]) 0.150683903
TensorDict({'a': a, ('b', 'b1'): b}, [3, 4]) 0.3777100720000002
TensorDict({'a': a, 'b': {'b1': b}}, [3, 4]) 0.27854125200000013
tdc = td.clone() 0.3433970380000009
tdc = td.clone();tdc['c'] = c 0.4057798770000005
tdc = td.clone();tdc.set('c', c) 0.39146732800000095
tdc = td.clone();tdc.share_memory_() 2.3296954949999993
tdc = td.clone();tdc.update(td2) 0.5400022369999995
tdc = td.clone();tdc.update(td2) 0.8455282950000012
tdc = td.clone();tdc['b', 'b1'] = b 0.5180912430000006
tdc = td.clone();tdc['c', 'c', 'c'] = c 0.8611584510000014
```

### CODE REVIEW

**Review by tcbegley — COMMENTED** [review]
This is great! Really simplifies a lot of things apart from anything else.

Perhaps we could add a test for `td._entry_type(key) == type(td.get(key))` and test all the different variants of `TensorDict`?

I left a few more comments inline.

File: tensordict/tensordict.py [hunk 1]
```diff
@@ -2659,6 +2535,9 @@ def _stack_onto_(
         torch.stack(list_item, dim=dim, out=self.get(key))
         return self
 
+    def _entry_class(self, key: Union[str, Tuple]) -> type:
+        return type(self.get(key))
```
[the remark this answers is not in the document]
Comment (vmoens) [reply tensordict/tensordict.py:2539]: exactly that

File: tensordict/tensordict.py [hunk 2]
```diff
@@ -261,19 +270,13 @@ def __new__(cls, *args, **kwargs):
...
-        state["_dict_meta"] = KeyDependentDefaultDict(self._make_meta)
         self.__dict__.update(state)
 
     def __init__(self):
-        self._dict_meta = KeyDependentDefaultDict(self._make_meta)
-
-    @abc.abstractmethod
-    def _make_meta(self, key: str) -> MetaTensor:
-        raise NotImplementedError
+        pass
```
[the remark this answers is not in the document]
Comment (vmoens) [reply tensordict/tensordict.py:279]: done

File: tensordict/tensordict.py [hunk 3]
```diff
@@ -108,6 +104,14 @@ class KeyedJaggedTensor:
 _STR_MIXED_INDEX_ERROR = "Received a mixed string-non string index. Only string-only or string-free indices are supported."
 
 
+def is_tensordict(datatype: type) -> bool:
+    return issubclass(datatype, TensorDictBase)
+
+
+def is_memmap(datatype: type) -> bool:
+    return issubclass(datatype, MemmapTensor)
```
[the remark this answers is not in the document]
Comment (vmoens) [reply tensordict/tensordict.py:112]: done

File: tensordict/tensordict.py [hunk 4]
```diff
@@ -108,6 +104,14 @@ class KeyedJaggedTensor:
 _STR_MIXED_INDEX_ERROR = "Received a mixed string-non string index. Only string-only or string-free indices are supported."
 
 
+def is_tensordict(datatype: type) -> bool:
+    return issubclass(datatype, TensorDictBase)
```
[the remark this answers is not in the document]
Comment (vmoens) [reply tensordict/tensordict.py:108]: done

**Pull request merged**